### PR TITLE
Adds a callback interface for Firefox Accounts updated profile

### LINKED ISSUE #5328: Introduce callbacks for an updated profile in firefox accounts
We currently cache profile data in our state, see https://github.com/mozilla/application-services/blob/280d69389689a17fac7b9d324338e83fec447fc1/components/fxa-client/src/internal/state_persistence.rs#L108

We also give the cache a 2-minute TTL, in order to not overwhelm the FxA server. See https://github.com/mozilla/application-services/blob/280d69389689a17fac7b9d324338e83fec447fc1/components/fxa-client/src/internal/profile.rs#L41

However, in the case our consumer has no network connection, the consumer would retrieve from cache for 2 minutes and thus getting last cached profile data, before suddenly experiencing no profile data. The reason is:

* When first running the app, the `get_profile` API correctly recognizes that it has been less than 2 minutes since the profile data was last cached, and thus returns from cache. So we see profile data
* After two minutes, the `get_profile` API would invalidate the cache, and attempt a network request, which would fail with an error. This leads the app to now think there is no profile, when there might be one

If we'd like to fix this, we can introduce a new `refresh_profile` API, that works hand in hand with a new callback interface to notify the app of an updated profile. The `refresh_profile` wouldn't return anything and instead just trigger the callback when:

* it reads the cache
* if it went to the network, when it receives a new value from the network

#5333 is a demo, but instead of re-using the existing "get_profile" function, we should introduce a new `refresh_profile` function (or `update_profile_if_needed` maybe a better name)

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/SYNC-3562)
┆Epic: Moving fxa-client functionality out of android-components

### PR DESCRIPTION
**Pull request by tarikeshaq**
This is a follow-up to the conversation on #5328. The idea is to add callbacks to profile fetching, where the app can register what happens when a profile is available and the Rust code would trigger that first on the cache, then on the value it gets from the FxA server if the `get_profile` call results in a server call.

Still a demo, but I'll add the kotlin/swift side of things in AS so it can be tested in the apps.

cc @mavduevskiy since I can't add you for review here (I think you may not be a part of the @mozilla github team)

Happy to take comments and pair on this too, I'll come back to it next week!

fixes #5328

### CODE REVIEW

**Comment by tarikeshaq** [comment]
I do want to point out that given the availability of callback, there are tons of other opportunities to refactor how the FxA client works - but it would be a significant amount of work to do that. This a small step in that direction

**Comment by tarikeshaq** [comment]
I created https://github.com/tarikeshaq/firefox-android/pull/1/files to show what consuming this could look like. The threading here is delicate, the consuming language should be careful of that when registering the callback. The callback is blocking from Rust's perspective, so if the consumer wants to do further I/O, the consumer should launch an async job and end the callback quickly

**Comment by mhammond** [comment]
This is fantastic, thanks for taking this on! Not quite sure about using the term "event" here, "callback" doesn't carry as much baggage but still sounds somehow worse - happy to let you and the Android team agree on terminology though.

**Review by jonalmeida — APPROVED** [review]
Thanks again for this excellent solution! I left one comment for myself and @mavduevskiy to consider.

File: components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/PersistedFirefoxAccount.kt [hunk 1]
```diff
@@ -516,6 +516,14 @@ class PersistedFirefoxAccount(inner: FirefoxAccount, persistCallback: PersistCal
         return this.inner.gatherTelemetry()
     }
 
+    fun registerEventHandler(handler: FirefoxAccountEventHandler) {
+        this.inner.registerEventHandler(handler)
+    }
+
+    fun unregisterEventHandler() {
```
[the remark this answers is not in the document]
Comment (mavduevskiy) [reply components/fxa-client/android/src/main/java/mozilla/appservices/fxaclient/PersistedFirefoxAccount.kt:523]: Still about to take a closer look, but stumbled upon [this bug](https://github.com/mozilla-mobile/android-components/issues/5104). Cache, auth problems, should we take it into consideration as well?

**Comment by tarikeshaq** [comment]
Pushed a commit that changes the approach so that the consumer passes the callback with the call to `refresh_profile` It's a little cleaner and this way the Rust layer doesn't store the callback - but the tradeoff is that we create a new instance of the callback every call (which I think is OK)

I'm ok with either approach and can revert the commit as well!

Pushed a commit to https://github.com/tarikeshaq/firefox-android/pull/1/files as well to show what consuming it looks like

**Comment by tarikeshaq** [comment]
Ah I think I'll close this. Ben is working on re-working some of the FxA APIs in #5609 and this hasn't gotten traction. We can re-open it if it's a priority

**Pull request closed without merging**